src/audiobuffer/audiobuffer.cpp

```diff
@@ -0,0 +1,252 @@
+
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <mutex>
+
+#include "audiobuffer.h"
+
+#ifdef __EMSCRIPTEN__
+#include <emscripten/emscripten.h>
+#endif
+
+// TODO: readSamplesFromBuffer as for waveform
+
+namespace SoLoud
+{
+	BufferStreamInstance::BufferStreamInstance(BufferStream *aParent)
+	{
+		mParent = aParent;
+		mOffset = 0;
+	}
+
+	BufferStreamInstance::~BufferStreamInstance()
+	{
+	}
+
+	unsigned int BufferStreamInstance::getAudio(float *aBuffer, unsigned int aSamplesToRead, unsigned int aBufferSize)
+	{
+		if (mParent->mBuffer.getFloatsBufferSize() == 0)
+		{
+			memset(aBuffer, 0, sizeof(float) * aSamplesToRead);
+			return 0;
+		}
+
+		unsigned int bufferSize = mParent->mBuffer.getFloatsBufferSize();
+		float *buffer = reinterpret_cast<float *>(mParent->mBuffer.buffer.data());
+		int samplesToRead = mOffset + aSamplesToRead > bufferSize ? bufferSize - mOffset : aSamplesToRead;
+		if (samplesToRead <= 0)
+			return 0;
+
+		if (samplesToRead != aSamplesToRead)
+		{
+			memset(aBuffer, 0, sizeof(float) * aSamplesToRead);
+
+			if (!mParent->dataIsEnded)
+			{
+				// The time the buffering is started.
+				// SoLoud::time currBufferTime = 1.0f * mOffset /
+				// 						  mParent->mPCMformat.bytesPerSample /
+				// 						  mParent->mChannels /
+				// 						  mParent->mPCMformat.sampleRate;
+				// std::cout << "CPP Buffering" << std::endl;
+			}
+		}
+
+		if (mChannels == 1)
+		{
+			// Optimization: if we have a mono audio source, we can just copy all the data in one go.
+			memcpy(aBuffer, buffer + mOffset, sizeof(float) * samplesToRead);
+		}
+		else
+		{
+			// From SoLoud documentation:
+			// So, if 1024 samples are requested from a stereo audio source, the first 1024 floats
+			// should be for the first channel, and the next 1024 samples should be for the second channel.
+			unsigned int i, j;
+			for (i = 0; i < samplesToRead; i++)
+			{
+				for (j = 0; j < mChannels; j++)
+				{
+					aBuffer[j * samplesToRead + i] = buffer[mOffset + i * mChannels + j];
+				}
+			}
```

Comment:
[nit, probably premature optimization] Maybe, for data locality sake, it would be better to have the inner loop outside and the outer loop inside. So we go from byte to neighboring byte instead of jumping between distant channel data.

Reply:
Nice catch!
Done.